libyul/optimiser/DeadCodeEliminator.cpp

```diff
@@ -0,0 +1,134 @@
+/*(
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that removes unreachable code.
+ */
+
+#include <libyul/optimiser/DeadCodeEliminator.h>
+#include <libyul/AsmData.h>
+
+#include <libevmasm/SemanticInformation.h>
+#include <libevmasm/AssemblyItem.h>
+
+#include <algorithm>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+using SemInfo = eth::SemanticInformation;
+
+namespace
+{
+bool isTerminating(yul::ExpressionStatement const& _exprStmnt)
+{
+	if (_exprStmnt.expression.type() != typeid(FunctionalInstruction))
+		return false;
+
+	auto funcInstr = boost::get<FunctionalInstruction>(_exprStmnt.expression);
+
+	return SemInfo::terminatesControlFlow(eth::AssemblyItem{funcInstr.instruction});
+}
+
+// Returns an iterator to the first terminating statement or ::end() when none
```

Comment:
```suggestion
// Returns an iterator to the first terminating statement or `_block.statements.cend()` when none
```

This is so dangerous. I really hate iterators ;)

Reply:
your correction is not exactly correct as it can be `cend()` or `end()` depending on whether the argument is const or not